code/modules/reagents/reagent_dispenser.dm

```diff
@@ -1,196 +1,340 @@
-	. = ..()
-	if(. && obj_integrity > 0)
-		if(tank_volume && (damage_flag == "bullet" || damage_flag == "laser"))
-			boom()
+/obj/structure/reagent_dispensers/on_reagent_change()
+	update_icon()
 
-/obj/structure/reagent_dispensers/attackby(obj/item/weapon/W, mob/user, params)
-	if(istype(W, /obj/item/weapon/screwdriver))
-		if(src.icon_state == "water" && !src.is_open_container())
-			src.icon_state = "wateropen"
-			user << "<span class='notice'>You unscrew and remove the lid from [src].</span>"
-			src.container_type |= OPENCONTAINER
-			src.update_icon()
-			playsound(src, 'sound/items/Deconstruct.ogg', 20, 1)
-			desc = "A water tank. It's top has been removed."
-			return
-		if(src.icon_state == "wateropen" && src.is_open_container())
-			src.icon_state = "water"
-			user << "<span class='notice'>You reattach and screw the [src]'s lid into place.</span>"
-			src.container_type &= ~OPENCONTAINER
-			src.update_icon()
-			playsound(src, 'sound/items/Deconstruct.ogg', 20, 1)
-			desc = "A water tank. It has a top that can be removed."
-			return
-		else
-			user << "<span class='warning'>This device does not have a removable lid!</span>"
-			return
-	if(istype(W, /obj/item/weapon/reagent_containers))
-		return 0 //so we can refill them via their afterattack.
-	else
-		return ..()
-
-/obj/structure/reagent_dispensers/New()
-	create_reagents(tank_volume)
-	reagents.add_reagent(reagent_id, tank_volume)
+/obj/structure/reagent_dispensers/attack_hand()
 	..()
+	update_icon()
 
-/obj/structure/reagent_dispensers/examine(mob/user)
-	..()
+/obj/structure/reagent_dispensers/update_icon()
+	overlays.Cut()
+	if(icon_state in bypassUIicons) return // can't do overlay fun things on stuff like peppertank/virus tank/watercooler
 	if(reagents.total_volume)
-		user << "<span class='notice'>It has [reagents.total_volume] units left.</span>"
-	else
-		user << "<span class='danger'>It's empty.</span>"
-
+		switch(reagents.get_master_reagent_id())//is master chem weldingfuel?change the iconstate into weldingfuel tank,same for water
+			if("welding_fuel")
+				name = "fueltank"
+				desc = "It can hold large amount of chemicals. Mainly used to hold fuel."
+				icon_state = "weldtank"
+			if("water")
+				name = "watertank"
+				desc = "It can hold large amount of chemicals. Mainly used to hold water."
+				icon_state = "watertank"
+			else
+				name = "storage tank"
+				desc = "It can hold large amount of chemicals."
+				icon_state = "chemtank"
+		var/image/filling = image('icons/obj/chemtank.dmi', src, "tankfilling1")
+		var/percent = round((reagents.total_volume / reagents.maximum_volume) * 100)
+		switch(percent)
+			if(0 to 25)			filling.icon_state = "tankfilling1"
+			if(25 to 50)		filling.icon_state = "tankfilling2"
+			if(50 to 75)		filling.icon_state = "tankfilling3"
+			if(75 to INFINITY)	filling.icon_state = "tankfilling4"
+		filling.color = mix_color_from_reagents(reagents.reagent_list)
+		overlays += filling

```

Comment:
**_Please do not manually change the overlays_**

Use `/proc/cut_overlays()` and `/proc/add_overlay(image, priority = 0)` so that overlay changes do not cause any lag